src/runtime/ndarray.cc

```diff
@@ -206,6 +214,41 @@ void NDArray::CopyFromTo(DGLArray* from, DGLArray* to) {
       from->dtype);
 }
 
+void NDArray::RecordedCopyFromTo(DGLArray* from, DGLArray* to, void* pyt_ctx) {
+  size_t from_size = GetDataSize(*from);
+  size_t to_size = GetDataSize(*to);
+  CHECK_EQ(from_size, to_size)
+      << "DGLArrayCopyFromTo: The size must exactly match.";
+
+  CHECK(from->ctx.device_type != to->ctx.device_type)
+      << "Recoding event is only called for the copy between CPU and GPU";
+
+  CHECK(from->ctx.device_type == kDGLCUDA || to->ctx.device_type == kDGLCUDA)
+      << "At least one CUDA ctx needs to be involved.";
+
+  DeviceAPI::Get(kDGLCUDA)->RecordedCopyDataFromTo(
+      from->data, static_cast<size_t>(from->byte_offset), to->data,
+      static_cast<size_t>(to->byte_offset), from_size, from->ctx, to->ctx,
+      from->dtype, pyt_ctx);
+}
+
+NDArray NDArray::PinnedEmpty(
+    std::vector<int64_t> shape, DGLDataType dtype, DGLContext ctx) {
+  CHECK_EQ(ctx.device_type, kDGLCPU) << "Only NDArray on CPU can be pinned";
+  NDArray ret = Internal::Create(shape, dtype, ctx);
+  // setup memory content (i.e., allocation)
+  size_t size = GetDataSize(ret.data_->dl_tensor);
+  if (size > 0) {
+    ret.data_->dl_tensor.data = DeviceAPI::Get(kDGLCUDA)->AllocPinnedDataSpace(
+        size, &(ret.data_->pyt_ctx), &(ret.data_->pyt_raw_deleter));
+    CHECK(
+        ret.data_->pyt_ctx != nullptr && ret.data_->pyt_raw_deleter != nullptr)
+        << "Failed to allocate memory from PyTorch's CUDACachingAllocator.";
```

Comment:
Is it possible to display the actual reason of the failure here?  Something like error code?

Reply:
I didn't find such info at pytorch side. I can add more info in this err msg, by printing the two ptrs, sth like 
```
<< "The allocation failed in PyTorch's CachingHostAllocator. "
<< "The returned context pointer is " << pyt_ctx_ 
<< " and the function deleter is " << pyt_raw_deleter_;

Reply:
I see, you can leave it.